crates/pallet-native-to-evm-currency-swap/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+
+        /// The pot native bridge account.
+        type PotNativeBridge: Get<Self::AccountId>;
+
+        /// The pot evm bridge account.
+        type PotEvmBridge: Get<Self::EvmAccountId>;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Balances were swapped.
+        BalancesSwapped {
+            /// The account id balances withdrawed from.
+            from: T::AccountId,
+            /// The withdrawed balances amount.
+            withdrawed_amount: NativeBalanceOf<T>,
+            /// The account id balances deposited to.
+            to: T::EvmAccountId,
+            /// The deposited balances amount.
+            deposited_amount: EvmBalanceOf<T>,
+            /// The corresponding transaction hash executed in evm.
+            evm_trx_hash: H256,
+        },
+    }
+
+    #[pallet::call(weight(<T as Config>::WeightInfo))]
+    impl<T: Config> Pallet<T> {
+        /// Swap balances.
+        #[pallet::call_index(0)]
+        pub fn swap(
+            origin: OriginFor<T>,
+            to: T::EvmAccountId,
+            #[pallet::compact] amount: NativeBalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            with_storage_layer(move || {
+                Self::do_swap(who, to, amount, Preservation::Expendable)?;
+
+                Ok(())
+            })
+        }
+
+        /// Same as the swap call, but with a check that the swap will not kill the origin account.
+        #[pallet::call_index(1)]
+        pub fn swap_keep_alive(
+            origin: OriginFor<T>,
+            to: T::EvmAccountId,
+            #[pallet::compact] amount: NativeBalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            with_storage_layer(move || {
+                Self::do_swap(who, to, amount, Preservation::Preserve)?;
+
+                Ok(())
+            })
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// General swap balances implementation.
+        pub fn do_swap(
+            who: T::AccountId,
+            to: T::EvmAccountId,
+            amount: NativeBalanceOf<T>,
+            preservation: Preservation,
+        ) -> DispatchResult {
+            let estimated_swapped_balance = T::BalanceConverter::convert(amount);
+            T::EvmCurrency::can_deposit(&to, estimated_swapped_balance, Provenance::Extant)
+                .into_result()?;
+
+            T::NativeCurrency::transfer(&who, &T::PotNativeBridge::get(), amount, preservation)?;
+
+            let evm_balance_to_be_deposited: u128 =
+                estimated_swapped_balance.unique_saturated_into();
+
+            let transaction = pallet_ethereum::Transaction::EIP1559(ethereum::EIP1559Transaction {
```

Comment:
Should we move transaction construction to runtime level and use some interface here instead ?

Reply:
No, why?

Reply:
Just a little question that I had in mind, I prefer to leave it here, but it's better to reask )

Reply:
Btw, can we use the `..Default::default()` notation here?